feat: add openfunction plugin (#7634)
apache/apisix: docs/en/latest/plugins/openfunction.md

---
title: openfunction
keywords:
  - APISIX
  - Plugin
  - OpenFunction
  - openfunction
description: This document contains information about the CNCF OpenFunction Plugin.
---

<!--
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
-->

## Description

> The `openfunction` Plugin is used to integrate APISIX with [CNCF OpenWhisk](https://openfunction.dev/) serverless platform.

Reviewer: ```suggestion
The `openfunction` Plugin is used to integrate APISIX with [CNCF OpenFunction](https://openfunction.dev/) serverless platform.
```

| ssl_verify        | boolean | False    | true    |              | When set to `true` verifies the SSL certificate.                                                           |
| service_token     | string  | False    |         |              | The token format is 'xx:xx' which support basic auth for ingress controller, .                                      |
| timeout           | integer | False    | 60000ms | [1, 60000]ms | OpenFunction action and HTTP call timeout in ms.                                                              |
| keepalive         | boolean | False    | true    |              | When set to `true` keeps the connection alive for reuse.                                                   |
| keepalive_timeout | integer | False    | 60000ms | [1000,...]ms | Time is ms for connection to remain idle without closing.                                                  |
| keepalive_pool    | integer | False    | 5       | [1,...]      | Maximum number of requests that can be sent on this connection before closing it.                          |

:::note

> The `timeout` attribute sets the time taken by the OpenFunction to execute, and the timeout for the HTTP client in APISIX. OpenFunction  calls may take time to pull the runtime image and start the container. So, if the value is set too small, it may cause a large number of requests to fail.

Reviewer 2: ```suggestion
The `timeout` attribute sets the time taken by the OpenFunction to execute, and the timeout for the HTTP client in APISIX. OpenFunction calls may take time to pull the runtime image and start the container. So, if the value is set too small, it may cause a large number of requests to fail.
```

:::

## Enabling the Plugin

> Before configuring the Plugin, you need to have OpenFunction running. The example below shows OpenFunction installed in Helm:

Reviewer: Also give a link to the installation page of OpenFunction.

> Before configuring the Plugin, you need to have OpenFunction running. For details, please refer to [Installation](https://openfunction.dev/docs/getting-started/installation/).

Reviewer 2: ```suggestion
Before configuring the plugin, you need to have OpenFunction running. For details, please refer to [Installation](https://openfunction.dev/docs/getting-started/installation/).
```

```shell
#add the OpenFunction chart repository
helm repo add openfunction https://openfunction.github.io/charts/
helm repo update

#install the OpenFunction chart
kubectl create namespace openfunction
helm install openfunction openfunction/openfunction -n openfunction
```

You can then verify if OpenFunction is ready:

```shell

> kubectl get pods -namespace openfunction

Reviewer: ```suggestion
kubectl get pods --namespace openfunction
```

---
title: openfunction
keywords:
  - APISIX
  - Plugin
  - OpenFunction
  - openfunction
description: This document contains information about the CNCF OpenFunction Plugin.
---

<!--
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
-->

## Description

The `openfunction` Plugin is used to integrate APISIX with [CNCF OpenFunction](https://openfunction.dev/) serverless platform.

> This Plugin can be configured on a Route and requests will be send to the configured OpenWhish API endpoint as the upstream.

Reviewer: It's not OpenWhish.

## Attributes

| Name                        | Type    | Required | Default | Valid values | Description                                                                                                |
| --------------------------- | ------- | -------- | ------- | ------------ | ---------------------------------------------------------------------------------------------------------- |
| function_uri                | string  | True     |         |              | function uri. For example, `https://localhost:30858/default/function-sample`.                              |
| ssl_verify                  | boolean | False    | true    |              | When set to `true` verifies the SSL certificate.                                                           |
| authorization               | object  | False    |         |              | Authorization credentials to access functions of OpenFunction.                                      |

> | authorization.service_token | string  | False    |         |              | The token format is 'xx:xx' which support basic auth for ingress controller, .                                      |

Reviewer: Why mention the ingress controller?

Author: It should be 'function entry points'

| keepalive                   | boolean | False    | true    |              | When set to `true` keeps the connection alive for reuse.                                                   |
| keepalive_timeout           | integer | False    | 60000ms | [1000,...]ms | Time is ms for connection to remain idle without closing.                                                  |
| keepalive_pool              | integer | False    | 5       | [1,...]      | Maximum number of requests that can be sent on this connection before closing it.                          |

:::note

The `timeout` attribute sets the time taken by the OpenFunction to execute, and the timeout for the HTTP client in APISIX. OpenFunction calls may take time to pull the runtime image and start the container. So, if the value is set too small, it may cause a large number of requests to fail.

:::

## Enabling the Plugin

Before configuring the plugin, you need to have OpenFunction running. For details, please refer to [Installation](https://openfunction.dev/docs/getting-started/installation/).
The example below shows OpenFunction installed in Helm:

```shell

> #add the OpenFunction chart repository

Reviewer: ```suggestion
# add the OpenFunction chart repository
```

Reviewer: Move these steps to the `Prerequisites` section (out of `Enabling the Plugin`).

Also, we need to tell users that Kubernetes is required.

helm repo add openfunction https://openfunction.github.io/charts/
helm repo update

#install the OpenFunction chart
kubectl create namespace openfunction
helm install openfunction openfunction/openfunction -n openfunction
```

You can then verify if OpenFunction is ready:

```shell

> kubectl get pods -namespace openfunction

Reviewer: ```suggestion
kubectl get pods --namespace openfunction
```

```

You can then create a function follow the [sample](https://github.com/OpenFunction/samples)

You can now configure the Plugin on a specific Route and point to this running OpenFunction service:

```shell
curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "uri": "/hello",
    "plugins": {
        "openfunction": {
            "function_uri": "http://localhost:3233/default/function-sample/test",
            "authorization": {
                "service_token": "test:test"
            }
        }
    }
}'
```

## Example usage

> Once you have configured the Plugin, you can send a request to the Route and it will invoke the configured function:

Reviewer: ```suggestion
Once you have configured the plugin, you can send a request to the Route and it will invoke the configured function:
```

| keepalive                   | boolean | False    | true    |              | When set to `true` keeps the connection alive for reuse.                                                   |
| keepalive_timeout           | integer | False    | 60000ms | [1000,...]ms | Time is ms for connection to remain idle without closing.                                                  |
| keepalive_pool              | integer | False    | 5       | [1,...]      | Maximum number of requests that can be sent on this connection before closing it.                          |

:::note

The `timeout` attribute sets the time taken by the OpenFunction to execute, and the timeout for the HTTP client in APISIX. OpenFunction calls may take time to pull the runtime image and start the container. So, if the value is set too small, it may cause a large number of requests to fail.

:::

## Prerequisites

Before configuring the plugin, you need to have OpenFunction running.
Installation of OpenFunction requires a certain version Kubernetes cluster.

> For details, please refer to [Installation](https://openfunction.dev/docs/getting-started/installation/).

Reviewer: Add a sub title:

```
### Install OpenFunction via Helm Chart
```

Author: Added

The example below shows OpenFunction installed in Helm:

```shell
# add the OpenFunction chart repository
helm repo add openfunction https://openfunction.github.io/charts/
helm repo update

#install the OpenFunction chart
kubectl create namespace openfunction
helm install openfunction openfunction/openfunction -n openfunction
```

You can then verify if OpenFunction is ready:

```shell
kubectl get pods --namespace openfunction
```

> 

Reviewer: Add a sub title:

```
### Push a Function
```

:::note

The `timeout` attribute sets the time taken by the OpenFunction to execute, and the timeout for the HTTP client in APISIX. OpenFunction calls may take time to pull the runtime image and start the container. So, if the value is set too small, it may cause a large number of requests to fail.

:::

## Prerequisites

Before configuring the plugin, you need to have OpenFunction running.
Installation of OpenFunction requires a certain version Kubernetes cluster.
For details, please refer to [Installation](https://openfunction.dev/docs/getting-started/installation/).

### Install OpenFunction via Helm Chart

The example below shows OpenFunction installed in Helm:

```shell
# add the OpenFunction chart repository
helm repo add openfunction https://openfunction.github.io/charts/
helm repo update

#install the OpenFunction chart
kubectl create namespace openfunction
helm install openfunction openfunction/openfunction -n openfunction
```

You can then verify if OpenFunction is ready:

```shell
kubectl get pods --namespace openfunction
```

> ### Create and push a function

Reviewer: ```suggestion
### Create and Push a Function
```

You can then create a function follow the [sample](https://github.com/OpenFunction/samples)

When building a function, you’ll need to push your function container image to a container registry like Docker Hub or Quay.io. To do that you’ll need to generate a secret for your container registry first.

```shell
REGISTRY_SERVER=https://index.docker.io/v1/ REGISTRY_USER=<your_registry_user> REGISTRY_PASSWORD=<your_registry_password>
kubectl create secret docker-registry push-secret \
    --docker-server=$REGISTRY_SERVER \
    --docker-username=$REGISTRY_USER \
    --docker-password=$REGISTRY_PASSWORD
```

> ## Enabling the Plugin

Reviewer: ```suggestion
## Enable the Plugin
```

:::

## Prerequisites

Before configuring the plugin, you need to have OpenFunction running.
Installation of OpenFunction requires a certain version Kubernetes cluster.
For details, please refer to [Installation](https://openfunction.dev/docs/getting-started/installation/).

### Create and Push a Function

You can then create a function following the [sample](https://github.com/OpenFunction/samples)

You'll need to push your function container image to a container registry like Docker Hub or Quay.io when building a function. To do that, you'll need to generate a secret for your container registry first.

```shell
REGISTRY_SERVER=https://index.docker.io/v1/ REGISTRY_USER= ${your_registry_user} REGISTRY_PASSWORD= ${your_registry_password}
kubectl create secret docker-registry push-secret \
    --docker-server=$REGISTRY_SERVER \
    --docker-username=$REGISTRY_USER \
    --docker-password=$REGISTRY_PASSWORD
```

## Enable the Plugin

You can now configure the Plugin on a specific Route and point to this running OpenFunction service:

```shell

> curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '

Reviewer 2: ```suggestion
curl http://127.0.0.1:9180/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
```
for the latest admin port change

---
title: openfunction
keywords:
  - APISIX
  - Plugin
  - OpenFunction
  - openfunction
description: This document contains information about the CNCF OpenFunction Plugin.
---

<!--
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
-->

## Description

The `openfunction` Plugin is used to integrate APISIX with [CNCF OpenFunction](https://openfunction.dev/) serverless platform.

> This Plugin can be configured on a Route and requests will be send to the configured OpenFunction API endpoint as the upstream.

Reviewer: ```suggestion
This Plugin can be configured on a Route and requests will be sent to the configured OpenFunction API endpoint as the upstream.
```

| keepalive                   | boolean | False    | true    |              | When set to `true` keeps the connection alive for reuse.                                                   |
| keepalive_timeout           | integer | False    | 60000ms | [1000,...]ms | Time is ms for connection to remain idle without closing.                                                  |
| keepalive_pool              | integer | False    | 5       | [1,...]      | Maximum number of requests that can be sent on this connection before closing it.                          |

:::note

The `timeout` attribute sets the time taken by the OpenFunction to execute, and the timeout for the HTTP client in APISIX. OpenFunction calls may take time to pull the runtime image and start the container. So, if the value is set too small, it may cause a large number of requests to fail.

:::

## Prerequisites

Before configuring the plugin, you need to have OpenFunction running.
Installation of OpenFunction requires a certain version Kubernetes cluster.
For details, please refer to [Installation](https://openfunction.dev/docs/getting-started/installation/).

> ### Install OpenFunction via Helm Chart

Reviewer 3: same as zh

Author: Thanks for your advice in docs!
docs updated!

The example below shows OpenFunction installed in Helm:

```shell
# add the OpenFunction chart repository
helm repo add openfunction https://openfunction.github.io/charts/
helm repo update

#install the OpenFunction chart
kubectl create namespace openfunction
helm install openfunction openfunction/openfunction -n openfunction
```

You can then verify if OpenFunction is ready:

```shell
kubectl get pods --namespace openfunction
```

### Create and Push a Function

You can then create a function follow the [sample](https://github.com/OpenFunction/samples)

> When building a function, you’ll need to push your function container image to a container registry like Docker Hub or Quay.io. To do that you’ll need to generate a secret for your container registry first.

Reviewer: "`suggestion
You'll need to push your function container image to a container registry like Docker Hub or Quay.io when building a function. To do that, you'll need to generate a secret for your container registry first.
```

```shell
REGISTRY_SERVER=https://index.docker.io/v1/ REGISTRY_USER=<your_registry_user> REGISTRY_PASSWORD=<your_registry_password>
kubectl create secret docker-registry push-secret \
    --docker-server=$REGISTRY_SERVER \
    --docker-username=$REGISTRY_USER \
    --docker-password=$REGISTRY_PASSWORD
```

## Enable the Plugin

You can now configure the Plugin on a specific Route and point to this running OpenFunction service:

```shell
curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "uri": "/hello",
    "plugins": {
        "openfunction": {
            "function_uri": "http://localhost:3233/default/function-sample/test",
            "authorization": {
                "service_token": "test:test"
            }
        }
    }
}'
```

## Example usage

Once you have configured the plugin, you can send a request to the Route and it will invoke the configured function:

```shell
curl -i http://127.0.0.1:9080/hello
```

This will give back the response from the function:

```
hello, test!
```

> ### Configuring path forwarding

Reviewer: ```suggestion
### Configure Path Transforming

```

> The `OpenFunction` Plugins also supports URL path forwarding while proxying requests to the OpenFunction API endpoints upstream. Extensions to the base request path gets appended to the `function_uri` specified in the Plugin configuration.

Reviewer: ```suggestion
The `OpenFunction` Plugin also supports transforming the URL path while proxying requests to the OpenFunction API endpoints. Extensions to the base request path get appended to the `function_uri` specified in the Plugin configuration.
```
